server/src/controllers/attainable.ts

```diff
@@ -233,3 +236,191 @@ export async function updateAttainable(req: Request, res: Response): Promise<voi
     }
   });
 }
+
+enum Status {
+  PASS = 'pass',
+  FAIL = 'fail',
+}
+type CalculationResult = {
+  status: Status;
+  points: number | undefined;
+}
+type WeightedAssignmentParams = {
+  min: number;
+  max: number;
+  weights: Array<number>;
+}
+
+type FormulaFunction = (subResults: Array<CalculationResult>) => Promise<CalculationResult>;
+type ParameterizedFormulaFunction = (parameters: any, subResults: Array<CalculationResult>) => Promise<CalculationResult>;
+type FormulaNode = {
+  validatedFormula: FormulaFunction;
+  subFormulaNodes: Array<FormulaNode>;
+};
+const formulasWithSchema: Map<Formula, [yup.AnyObjectSchema, ParameterizedFormulaFunction]> = new Map();
```

Comment:
How would you feel about `/types`? Or would this perhaps warrant a `/services` entry?

Reply:
The definition of formula related types is fine under `src/types`, but the formula functions themselves aren't exactly types. We discussed that we could have a new directory `src/formulas` where formulas are stored, since `services` sounds ambiguous. In the future there may also be a lot of formulas, so formulas having their own directory would be justified.